pipeline/src/pipeline.rs

```diff
@@ -827,6 +828,12 @@ impl<T: FieldElement> Pipeline<T> {
         Ok(self.artifact.witness.as_ref().unwrap().clone())
     }
 
+    pub fn publics(&self) -> Result<Vec<(String, T)>, Vec<String>> {
+        let pil = self.optimized_pil()?;
```

Comment:
Is `optimized_pil` the right method to call here? And should this method take a mutable reference and compute anything that isn't there yet?

Reply:
The `compute_*` methods take a `&mut self` and actually compute whatever is needed. `optimized_pil` takes an immutable ref an assumes all the needed deps have been computed.

But don't you need the witness here too?

Reply:
Yes I get it on the next line